go/ct/gen/transient.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package gen
+
+import (
+	"fmt"
+	"slices"
+	"strings"
+
+	"golang.org/x/exp/maps"
+	"pgregory.net/rand"
+
+	"github.com/Fantom-foundation/Tosca/go/ct/common"
+	"github.com/Fantom-foundation/Tosca/go/ct/st"
+)
+
+type TransientStorageGenerator struct {
+	nonZeroConstraints map[Variable]bool
+	unsatisfiable      bool
+}
+
+func NewTransientStorageGenerator() *TransientStorageGenerator {
+	return &TransientStorageGenerator{nonZeroConstraints: map[Variable]bool{}}
+}
+
+// BindToNonZero binds the transient storage location to a non-zero value.
+func (t *TransientStorageGenerator) BindToNonZero(key Variable) {
+	if value, exits := t.nonZeroConstraints[key]; exits {
+		if !value {
+			t.unsatisfiable = true
+			return
+		}
+	} else {
+		t.nonZeroConstraints[key] = true
+	}
+}
+
+// BindToZero binds the transient storage location to a zero value.
+func (t *TransientStorageGenerator) BindToZero(key Variable) {
+	if value, exits := t.nonZeroConstraints[key]; exits {
+		if value {
+			t.unsatisfiable = true
+			return
+		}
+	} else {
+		t.nonZeroConstraints[key] = false
+	}
+}
+
+func (t *TransientStorageGenerator) Generate(assignment Assignment, rnd *rand.Rand) (*st.TransientStorage, error) {
+	if t.unsatisfiable {
+		return nil, fmt.Errorf("%w, conflicting set/unset constraints", ErrUnsatisfiable)
+	}
+
+	keysInUse := map[common.U256]bool{}
+	for variable := range t.nonZeroConstraints {
+		if pos, isBound := assignment[variable]; isBound {
+			keysInUse[pos] = true
+		}
+	}
+
+	getUnusedKey := func() common.U256 {
+		for {
+			key := common.RandU256(rnd)
+			if _, isPresent := keysInUse[key]; !isPresent {
+				keysInUse[key] = true
+				return key
+			}
+		}
+	}
+
+	transient := &st.TransientStorage{}
+
+	// Process zero/nonZero constraints.
+	for variable, nonZero := range t.nonZeroConstraints {
+
+		// Check different assignments with the same key if they conflict
+		for other, otherNonZero := range t.nonZeroConstraints {
+			if assignment[variable] == assignment[other] && nonZero != otherNonZero {
+				return nil, fmt.Errorf("%w, conflicting constraints %v and %v on the same location", ErrUnsatisfiable, nonZero, otherNonZero)
+			}
+		}
+
+		var key common.U256
+		if pos, isBound := assignment[variable]; isBound {
+			key = pos
+		} else {
+			key = getUnusedKey()
+		}
+
+		if nonZero {
+			transient.Set(key, common.RandU256(rnd))
```

Comment:
bind to zero means that the key will not be in the storage (transient)? 
or does it mean that a 0 value is stored?  if so, shall we write an entry in the storage with 0 content?

Reply:
For transient storage it does not make a difference whether 0 is stored or not. See earlier comments on this PR where we decided that it is not required to be saved.